pkg/services/repo_service.go

```diff
@@ -66,3 +72,75 @@ func (a *argoCDService) GetApps(ctx context.Context, repoURL string, revision st
 
 	return res, nil
 }
+
+func (a *argoCDService) GetPaths(ctx context.Context, repoURL string, revision string, pattern string) ([]string, error) {
+	repo, err := a.repositoriesDB.GetRepository(ctx, repoURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error in GetRepository")
+	}
+
+	gitRepoClient, err := git.NewClient(repo.Repo, repo.GetGitCreds(), repo.IsInsecure(), repo.IsLFSEnabled())
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = checkoutRepo(gitRepoClient, revision)
+	if err != nil {
+		return nil, err
+	}
+
+	paths, err := gitRepoClient.LsFiles(pattern)
```

Comment:
The `LsFiles` implementation from Argo CD's git client does not handle recursive filters (eg`**/*.json` to recursively match all JSON files in a repo)

`LsFiles` is a wrapper around `git ls-files --full-name -- (filter)`, which treats `**/*.json` as equivalent to `*/*.json` (only matching from root folder)

Reply:
And related: realized that since Argo CD's `util/git` utility calls to `git` directly, `git` needs to be included in the [Dockerfile](https://github.com/argoproj-labs/applicationset/blob/master/Dockerfile)

Reply:
@jgwest I ran `git ls-files --full-name -- **/*.json` on https://github.com/rtnpro/argocd-applicationset-examples/ and it returned

```
cluster-config/engineering/local/config.json
```

So, it supports searching for matching files in nested directories.

Reply:
@jgwest I have added more `.json` files to the repo: https://github.com/rtnpro/argocd-applicationset-examples/ to validate the output of `git ls-files --full-name -- <pattern>` and it seems to work as expected.

![image](https://user-images.githubusercontent.com/231684/102905250-d8ec1380-4498-11eb-909a-c77492877248.png)

cc @OmerKahani

Reply:
@rtnpro Agreed, thanks for investigating, it appears to work as expected for me as well... I'm not able to reproduce with the information I originally provided, and I'm not able to reproduce now (and no longer have the environment I originally used),  so this can be resolved.